marathon/src/main/scala/io/buoyant/marathon/v2/Api.scala

```diff
@@ -101,21 +133,131 @@ object Api {
   }
 }
 
-private class AppIdApi(client: Api.Client, apiPrefix: String)
+private class AppIdApi(client: Api.Client, apiPrefix: String, auth: Option[Api.AuthRequest])
   extends Api
   with Closable {
 
   import Api._
 
   def close(deadline: Time) = client.close(deadline)
 
-  def getAppIds(): Future[Api.AppIds] = {
-    val req = http.Request(s"$apiPrefix/apps")
-    Trace.letClear(client(req)).flatMap(rspToApps(_))
+  private[this] val authTokenState = Var[Activity.State[Option[String]]](Activity.Ok(None))
+  private[this] val authToken = Activity(authTokenState)
+
+  private[this] def handleTokenResponse(rsp: Try[http.Response]): Future[String] = rsp match {
+    case Return(rsp) =>
+      rsp.status match {
+        case http.Status.Ok =>
+          readJson[AuthToken](rsp.content) match {
+            case Return(AuthToken(Some(token))) =>
+              authTokenState() = Activity.Ok(Some(token))
+              Future.value(token)
+            case Throw(e) =>
+              authTokenState() = Activity.Ok(None)
+              log.error(s"refreshToken threw $e")
+              Future.exception(e)
+            case _ =>
+              authTokenState() = Activity.Ok(None)
+              val e = UnexpectedResponse(rsp)
+              log.error(s"refreshToken returned Unexpected with $e")
+              Future.exception(e)
+          }
+        case http.Status.Unauthorized =>
+          authTokenState() = Activity.Ok(None)
+          val e = UnauthorizedResponse(rsp)
+          log.error(s"refreshToken returned Unauthorized with $e")
+          Future.exception(e)
+        case _ =>
+          authTokenState() = Activity.Ok(None)
+          val e = UnexpectedResponse(rsp)
+          log.error(s"refreshToken returned Unexpected with $e")
+          Future.exception(e)
+      }
+    case Throw(e) =>
+      authTokenState() = Activity.Ok(None)
+      log.error(s"refreshToken threw $e")
+      Future.exception(e)
   }
 
-  def getAddrs(app: Path): Future[Set[Address]] = {
-    val req = http.Request(s"$apiPrefix/apps${app.show}")
-    Trace.letClear(client(req)).flatMap(rspToAddrs(_))
+  // Given AuthRequest data, attempts authentication on a marathon login
+  // endpoint. Upon success, returns a short-lived authToken, and also saves
+  // authToken for subsequent API calls.
+  private[this] def refreshToken(): Future[String] =
+    auth match {
+      case None => Future.value("")
```

Comment:
I think this means we can issue a request with `Authorization: token=` -- is that ever appropriate?